x/circuit/keeper/msg_server.go

```diff
@@ -0,0 +1,219 @@
+		perms, err := srv.GetPermissions(ctx, address)
+		if err != nil {
+			return nil, err
+		}
+
+		if perms.Level != types.Permissions_LEVEL_SUPER_ADMIN {
+			return nil, fmt.Errorf("only super admins can authorize users")
+		}
+	}
+
+	grantee, err := srv.addressCodec.StringToBytes(msg.Grantee)
+	if err != nil {
+		return nil, err
+	}
+
+	// Append the account in the msg to the store's set of authorized super admins
+	if err = srv.SetPermissions(ctx, grantee, msg.Permissions); err != nil {
+		return nil, err
+	}
+
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			"authorize_circuit_breaker",
+			sdk.NewAttribute("granter", msg.Granter),
+			sdk.NewAttribute("grantee", msg.Grantee),
+			sdk.NewAttribute("permission", msg.Permissions.String()),
+		),
+	})
+
+	return &types.MsgAuthorizeCircuitBreakerResponse{
+		Success: true,
+	}, nil
+}
+
+func (srv msgServer) TripCircuitBreaker(goCtx context.Context, msg *types.MsgTripCircuitBreaker) (*types.MsgTripCircuitBreakerResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	address, err := srv.addressCodec.StringToBytes(msg.Authority)
+	if err != nil {
+		return nil, err
+	}
+
+	store := ctx.KVStore(srv.storekey)
+
+	// Check that the account has the permissions
+	perms, err := srv.GetPermissions(ctx, address)
+	if err != nil {
+		return nil, err
+	}
+
+	switch {
+	case perms.Level == types.Permissions_LEVEL_SUPER_ADMIN || bytes.Equal(address, srv.GetAuthority()):
+		// add all msg type urls to the disable list
+		for _, msgTypeUrl := range msg.MsgTypeUrls {
+			if !srv.IsAllowed(ctx, msgTypeUrl) {
+				return nil, fmt.Errorf("message %s is already disabled", msgTypeUrl)
+			}
+			store.Set(types.CreateDisableMsgPrefix(msgTypeUrl), []byte{0x01})
+		}
+	case perms.Level == types.Permissions_LEVEL_ALL_MSGS:
+		// iterate over the msg type urls
+		for _, msgTypeUrl := range msg.MsgTypeUrls {
+			// check if the message is in the list of allowed messages
+			if !srv.IsAllowed(ctx, msgTypeUrl) {
+				return nil, fmt.Errorf("message %s is already disabled", msgTypeUrl)
+			}
+			store.Set(types.CreateDisableMsgPrefix(msgTypeUrl), []byte{0x01})
+		}
+	case perms.Level == types.Permissions_LEVEL_SOME_MSGS:
+		for _, msgTypeUrl := range msg.MsgTypeUrls {
+			// check if the message is in the list of allowed messages
+			if !srv.IsAllowed(ctx, msgTypeUrl) {
+				return nil, fmt.Errorf("message %s is already disabled", msgTypeUrl)
+			}
+			for _, msgurl := range perms.LimitTypeUrls {
+				if msgTypeUrl == msgurl {
+					store.Set(types.CreateDisableMsgPrefix(msgTypeUrl), []byte{0x01})
+				} else {
+					return nil, fmt.Errorf("account does not have permission to trip circuit breaker for message %s", msgTypeUrl)
+				}
+			}
+		}
+	default:
+		return nil, fmt.Errorf("account does not have permission to trip circuit breaker")
+	}
+
+	var msg_urls string
+	if len(msg.GetMsgTypeUrls()) > 1 {
+
+		for _, url := range msg.GetMsgTypeUrls() {
```

Comment:
```suggestion
strings.Join(msg.GetMsgTypeUrls, ",")
```

Curious why it was separating the string with a space, was it intended?